src/app/AttributeAccessInterface.h

```diff
@@ -41,23 +42,143 @@
+     * Note: Only append is supported currently (encode a null list index), other operations won't encode a list index in the
+     * attribute path field.
+     */
+    CHIP_ERROR PrepareAttribute(AttributeReportIBs::Builder & aAttributeReportIBs, const ConcreteDataAttributePath & aPath,
+                                DataVersion aDataVersion);
+
+    /**
+     * FinishAttribute encodes the "footer" part of an attribute report (it closes the containers opened in PrepareAttribute)
+     */
+    CHIP_ERROR FinishAttribute();
+
+    template <typename... Ts>
+    CHIP_ERROR Encode(Ts... aArgs)
+    {
+        return DataModel::Encode(*mAttributeDataIBBuilder.GetWriter(), TLV::ContextTag(to_underlying(AttributeDataIB::Tag::kData)),
+                                 std::forward<Ts>(aArgs)...);
+    }
+
+private:
+    AttributeReportIB::Builder mAttributeReportIBBuilder;
+    AttributeDataIB::Builder mAttributeDataIBBuilder;
+};
+
+/**
+ * The AttributeValueEncoder is a helper class for filling report payloads into AttributeReportIBs.
+ * The attribute value encoder can be initialized with a AttributeEncodeState for saving and recovering its state between encode
+ * sessions (chunkings).
+ *
+ * When Encode returns recoverable errors (e.g. CHIP_ERROR_NO_MEMORY) the state can be used to initialize the AttributeValueEncoder
+ * for future use on the same attribute path.
+ */
+class AttributeValueEncoder
+{
+public:
+    class ListEncodeHelper
+    {
+    public:
+        ListEncodeHelper(AttributeValueEncoder & encoder) : mAttributeValueEncoder(&encoder) {}
+
+        template <typename... Ts>
+        CHIP_ERROR Encode(Ts... aArgs) const
+        {
+            return mAttributeValueEncoder->EncodeListItem(std::forward<Ts>(aArgs)...);
+        }
+
+    private:
+        AttributeValueEncoder * mAttributeValueEncoder;
+    };
+
+    class AttributeEncodeState
+    {
+    public:
+        AttributeEncodeState() : mAllowPartialData(false), mCurrentEncodingListIndex(kInvalidListIndex) {}
+        bool AllowPartialData() const { return mAllowPartialData; }
+
+    private:
+        friend class AttributeValueEncoder;
+
+        bool mAllowPartialData              = false;
+        ListIndex mCurrentEncodingListIndex = kInvalidListIndex;
+    };
+
+    /**
+     * AttributeReportBuilder offleads some code in AttributeValueEncoder::Encode to keep its content small.
+     */
+    AttributeValueEncoder(AttributeReportIBs::Builder & aAttributeReportIBsBuilder, FabricIndex aAccessingFabricIndex,
+                          const ConcreteAttributePath & aPath, DataVersion aDataVersion,
+                          const AttributeEncodeState & aState = AttributeEncodeState()) :
+        mAttributeReportIBsBuilder(aAttributeReportIBsBuilder),
+        mAccessingFabricIndex(aAccessingFabricIndex), mPath(aPath), mDataVersion(aDataVersion), mEncodeState(aState)
     {}
 
     template <typename... Ts>
     CHIP_ERROR Encode(Ts... aArgs)
```

Comment:
So what are the semantics of this function now?  Before they were pretty straightforward: you hand it the entire attribute value, and it gets encoded.  Now it's doing ... something else.  What is it doing?  This needs to be very clearly documented, so consumers have some hope of using this correctly.

In particular, I would have expected that the only ListOperation values we can get out of here are either NotList or ReplaceAll, since this represents encoding the whole attribute value.  But clearly that's not the setup anymore.  What is the setup?

Reply:
Added a comment, and more comments to the AttributeReportBuilder class & EncodeList function.
The current implementation is also straightforward (for atomic type) -- it encodes a single attribute report IB for an atomic attribute value.